lib/namespace-handler-hmac.js

```diff
@@ -0,0 +1,174 @@
+              self.getKey(options, done);
+            });
+          }
+          callback(null, state);
+        });
+      }],
+      getKey: ['rotate', (callback, results) => {
+        callback(null, {
+          id: results.rotate.key.id,
+          material: new Buffer(results.rotate.key.data, 'base64')
+        });
+      }]
+    }, (err, results) => done(err, results ? results.getKey : null));
+  }
+
+  /**
+   * Verify a JWT.
+   *
+   * @param token the token to be verified.
+   * @param options the options to use:
+   *   namespace the state to verify with.
+   *   keyId the key ID to use.
+   *
+   * @param callback(err, tokenPayload) called once the operation completes.
+   */
+  verify(token, options, callback) {
+    // get key that matches key ID
+    const state = options.namespace.state;
+    let key;
+    if(options.keyId === state.key.id) {
+      key = state.key;
+    } else if(state.previousKey && options.keyId === state.previousKey.id) {
+      key = state.previousKey;
+    } else {
+      return callback(new Error('Invalid key identifier in token.'));
+    }
+    // decode and verify the token
+    jwt.verify(
+      token, new Buffer(key.data, 'base64'), {
+        clockTolerance: options.namespace.clockToleranceInSecs
+      }, callback);
+  }
+
+  _get(id, callback) {
+    database.collections.jwtKeyStore.findOne(
+      {id: database.hash(id)}, {namespace: true}, (err, record) => {
+        if(err) {
+          logger.error('Mongo error when trying to find key state.', err);
+          return callback(new BedrockError(
+            'Failed to find key state due to internal error.',
+            'InternalError',
+            {key: id, httpStatusCode: 500, 'public': true}, err));
+        }
+        if(!record) {
+          return callback(new BedrockError(
+            'Key not found.',
+            'NotFound',
+            {key: id, httpStatusCode: 404, 'public': true}));
+        }
+        callback(null, record.namespace);
+      });
+  }
+
+  _update(id, state, options, callback) {
+    if(typeof options === 'function') {
+      callback = options;
+      options = {};
+    }
+    options = options || {};
+    const query = _.assign({}, options.query || {});
+    query.id = database.hash(id);
+    database.collections.jwtKeyStore.update(
+      query, {
+        $set: {
+          'meta.updated': Date.now(),
+          'namespace.state': state
+        }
+      }, database.writeOptions, (err, result) => {
+        if(err) {
+          logger.error('Mongo error when trying to update key state.', err);
+          return callback(new BedrockError(
+            'Failed to update key state due to internal error.',
+            'InternalError',
+            {namespace: id, httpStatusCode: 500, 'public': true},
+            err));
+        }
+        callback(null, result.result.n > 0);
+      });
+  }
+
+  _generateHmacKey() {
+    // TODO: 16 bytes is based on HS256, could be different
```

Comment:
Remove TODO ... I put this here and looking back now it actually makes no sense. 16 bytes is fine, as 128 bits is ridiculously difficult to crack. it's the hash that is 256 bits and this is unrelated. We could make it 32 bits and just never think about it again until someone can harness a few supernovas to crack it. In other words, just remove the TODO. :)